dbcon/execplan/rewrites.cpp

```diff
@@ -110,44 +114,104 @@ bool simpleFiltersCmp(const SimpleFilter* left, const SimpleFilter* right)
 }
 
 // Walk the tree and find out common conjuctions
-void collectCommonConjuctions(execplan::ParseTree* root, CommonContainer& accumulator, int level = 0,
-                              bool orMeeted = false, bool andParent = false)
+struct StackFrameWithSet
 {
-  if (root == nullptr)
+  execplan::ParseTree* node;
+  GoTo direction;
+  bool orMet;
+  bool andParent;
+  CommonContainer localset;
+  StackFrameWithSet(execplan::ParseTree* node_, GoTo direction_, bool orMet_ = false, bool andParent_ = false)
+   : node(node_), direction(direction_), orMet(orMet_), andParent(andParent_), localset({{}, {}})
   {
-    return;
   }
+};
 
-  printTreeLevel(root, level);
-  // the condition below means leaf node
-  if (root->left() == nullptr && root->right() == nullptr && orMeeted && andParent)
+void advanceSetUp(std::vector<StackFrameWithSet>& stack, CommonContainer& accumulator)
+{
+  if (stack.size() == 1)
+    accumulator = stack.back().localset;
+  else
   {
-    // we want to collect it if it is a child of and node and or node was met before
-    if (castToFilter(root))
+    auto sz = stack.size();
+    if (operatorType(stack.at(sz - 2).node) == OP_OR)
```

Comment:
Do we check the invariant that there is something at(sz-2)? If not std::optional looks like a natural fit here.

Reply:
stack size is always > 0  in the while loop outside. and the = 1 case is checked on line 132.